src/drivers/imu/bno055/bno055.cpp

```diff
@@ -0,0 +1,197 @@
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "bno055.hpp"
+
+using namespace std::placeholders;
+
+BNO055::BNO055(I2CSPIBusOption bus_option, int bus, int bus_frequency) :
+	I2C(DRV_IMU_DEVTYPE_BNO055, MODULE_NAME, bus, BNO055_I2C_ADDR1, bus_frequency),
+	I2CSPIDriver(MODULE_NAME, px4::device_bus_to_wq(get_device_id()), bus_option, bus),
+	_px4_accel(get_device_id()),
+	_px4_gyro(get_device_id()),
+	_px4_mag(get_device_id())
+{
+}
+
+BNO055::~BNO055()
+{
+}
+
+int BNO055::init()
+{
+	// do I2C init
+	int ret = I2C::init();
+
+	if (ret != OK) {
+		PX4_ERR("I2C setup failed");
+		return ret;
+	}
+
+	// set up bno lib
+	// the bind() is used since the (C) lib wants pointer-to-function, but our (C++) wrapper has pointer-to-member
+	bno055_struct.bus_write = &BNO055::write_reg_trampoline;
+	bno055_struct.bus_read = &BNO055::read_reg_trampoline;
+	bno055_struct.delay_msec = &BNO055::bno_msleep;
+	bno055_struct.dev_addr = BNO055_I2C_ADDR1; // TODO
+	bno055_struct.i2c_calling_obj = this;
+
+	// init the bosch lib
+	ret = bno055_init(&bno055_struct);
+
+	if (ret != OK) {
+		PX4_ERR("Bosch lib init failed");
+		return ret;
+	}
+
+	// check if the sensor has the right chip id (read in bno055_init)
+	if (bno055_struct.chip_id != 0xA0) {
+		PX4_ERR("Chip ID is wrong, need A0, got %x", bno055_struct.chip_id);
+		return -1;
+	}
+
+	// reset sensor first
+	ret = reset();
+
+	if (ret != OK) {
+		PX4_ERR("Can't reset sensor");
+		return ret;
+	}
+
+	// wait some time for the sensor to reset
+	bno_msleep(700);
```

Comment:
This is executed within a work queue per bus so we try to avoid sleeps within drivers because you'll mess up the scheduling of anything else on that bus.

Most of our drivers have a simple internal state machine to handle things like this so they can request the reset and come back later to check if it's complete.
https://github.com/PX4/PX4-Autopilot/blob/de6661634161be00c5fd7949875b401e71f21358/src/drivers/imu/bosch/bmi088/BMI088_Accelerometer.cpp#L115-L122

Reply:
Even in the init() function? As far as I know there is no sleeping in the RunImpl(). The only problem is that I want the sensor to be in a known state, and it needs some time to execute the reset / mode switches.